Add Navigation block toolbar buttons to Template Part editor

getdave opened the pull request:
## What

Adds Navigation block toolbar buttons to Template Part editor for improved discoverability and navigation editing workflow.

- Adds 'Edit navigation' button when template part contains Navigation blocks
- Button selects the first Navigation block and opens inspector for List View editing
- Prevents double-click navigation when editing Navigation blocks
- Works in all editing modes (not restricted to contentOnly)

<img width="1798" height="371" alt="Screen Shot 2025-10-01 at 17 24 29" src="https://github.com/user-attachments/assets/ca313e48-1040-4d71-95d2-00255fdce827" />

Extracted from #71137 with contentOnly restrictions removed.

Closes https://github.com/WordPress/gutenberg/issues/72009

## Why

This improves discoverability of Navigation block editing within Template Parts by providing direct access through the toolbar, regardless of editing mode. 

In recent user testing sessions, I have observed users becoming confused about where to edit Navigation. As the template part has an initial overlay it's useful to show this additional prompt to drop them into the editing of the menu.

The original PR #71137 was restricted to contentOnly mode, but this functionality is valuable across all editing contexts.

## How

- Enhanced Template Part editor with Navigation block detection logic
- Added conditional 'Edit navigation' toolbar button
- Added visual separation between toolbar button groups using different BlockControls groups
- Modified double-click navigation behavior to respect Navigation editing context
- Removed contentOnly mode restrictions for broader applicability

## Testing

1. Add a Navigation block to a Template Part
2. Verify 'Edit navigation' button appears in Template Part toolbar
3. Click button to verify Navigation block selection and inspector opening
4. Test in both default and contentOnly editing modes
5. Verify double-click navigation is disabled when editing Navigation blocks

## Screenshots



https://github.com/user-attachments/assets/9b7d0a16-9254-4489-98c9-a684a1e090d1







## Uncertain Elements

- This PR contains prototype hacks from the original implementation that would need proper APIs for production
- The visual separation using different BlockControls groups is a workaround
- Navigation context detection uses hard-coded block name comparisons

getdave committed:
Add content role to Navigation Link and Submenu label attributes

This change makes navigation menu item labels editable in Write Mode by adding role: 'content' to the label attribute for both core/navigation-link and core/navigation-submenu blocks.

This allows users to:
- Edit menu item text/labels in Write Mode
- Maintain the content-focused editing experience
- Keep layout and styling controls disabled in Write Mode

Part of the Navigation Block Write Mode improvements.

jasmussen commented:
Looks good, feels valid. If we go this way, one change:

<img width="311" height="61" alt="image" src="https://github.com/user-attachments/assets/6fd45963-b23a-4e8a-ad00-23eafbd988d9" />

I would add a vertical separator before Edit header as well. This is in part because the group makes it feel more like a separate button, but also conceptually the button is not related to the alignment button.

jeryj commented on packages/block-library/src/template-part/edit/index.js:
```diff
@@ -247,10 +285,36 @@ export default function TemplatePartEdit( {
 								}
 							>
-								{ __( 'Edit' ) }
+								{ hasNavigationBlocks
+									? sprintf(
+											/* translators: %s: template part title or fallback text */
+											__( 'Edit %s' ),
+											(
+												title || __( 'template part' )
+											).toLowerCase()
+									  )
+									: __( 'Edit' ) }
```
I could see the `Edit %s` being helpful even if there isn't a navigation. It's certainly clearer. That way when you're looking to edit a template, it's always the same regardless if there's a navigation block inside the template or not. It's more consistent.

jeryj commented on packages/block-library/src/template-part/edit/index.js:
```diff
@@ -232,13 +264,19 @@ export default function TemplatePartEdit( {
 		);
 	}
 
+	/*
+	 * PROTOTYPE HACK: Using different group props to create visual separation
+	 * between toolbar buttons. WordPress automatically adds separators between
+	 * different BlockControls groups. This avoids the need for custom CSS
+	 * while maintaining the standard toolbar appearance.
+	 */
```
This is done elsewhere in the codebase and the groups don't assign any kind of meaningful descriptions or semantics, so this is fine to me. I don't know what the PROTOTYPE HACK is about though. Is this not meant to be merged?

getdave replied:
Not meant to be merged. I think my cherry picking went very wrong!

I spent time yesterday hoping I'd remlved all these unwanted changes 🤦

jeryj commented on packages/block-library/src/template-part/edit/index.js:
```diff
@@ -247,10 +285,36 @@ export default function TemplatePartEdit( {
+								// Enable the complementary area (inspector)
+								enableComplementaryArea(
+									'core',
+									'edit-post/block'
```
Is `edit-post/block` always the right complimentary area?

getdave replied:
Yeh I'm not sure on that one 🤔

getdave replied:
I believe it is yes

getdave commented on packages/block-library/src/template-part/edit/index.js:
```diff
@@ -232,13 +264,19 @@ export default function TemplatePartEdit( {
 		);
 	}
 
+	/*
+	 * PROTOTYPE HACK: Using different group props to create visual separation
+	 * between toolbar buttons. WordPress automatically adds separators between
+	 * different BlockControls groups. This avoids the need for custom CSS
+	 * while maintaining the standard toolbar appearance.
+	 */
```
```suggestion
```

Remove

getdave committed:
Remove unwanted comments

getdave committed:
Refactor: Move Edit Navigation button to editor hook

Refactors the "Edit Navigation" button for Template Parts from hardcoded logic in the block-library component to a reusable editor hook following the same pattern as navigation-link-view-button.

Changes:
- Created template-part-navigation-button hook in editor package
- Removed navigation-specific logic from Template Part edit component
- Added flashBlock with 500ms duration for visual feedback
- Hook automatically opens block inspector when button is clicked

This removes the interfaceStore dependency from block-library and follows established patterns for extending block toolbar functionality.

getdave committed:
Improve naming

getdave committed:
Add divider

getdave commented on packages/editor/src/hooks/template-part-navigation-edit-button.js:
```diff
@@ -0,0 +1,134 @@
+	const { selectBlock, flashBlock } = useDispatch( blockEditorStore );
+	const { enableComplementaryArea } = useDispatch( interfaceStore );
+
+	const { hasNavigationBlocks, firstNavigationBlockId } = useSelect(
```
I'm looking at perf optimisation now.

getdave committed:
Address review comments

getdave committed:
optimize perf

jeryj committed:
Don't show edit navigation button if navigation block is in disabled editing mode

jeryj reviewed (approved):
Tested this out manually and it matches your video. I did find that it didn't work from wp-admin -> edit page -> device dropdown -> show template. From that context, you can only edit in isolation mode, so clicking the Edit navigation button selected the disabled navigation block without any content editing options. 

[I added a check to see if the navigation block is disabled](https://github.com/WordPress/gutenberg/pull/72008/commits/9fb5712f034b54d313ce85b5ba0f9592a63be990), and only showing the Edit navigation button if it's not disabled, resolved the issue. Hopefully we will merge the editing experience between the standalone page editor and page editor within the site editor. That's definitely outside of the scope of this PR though :)
